File: Packages/tongue-js/src/hashing.ts
// FNV-1a character n-gram featurizer, ported from src/tongue_training/hashing.py.
//
// Hashing instead of a learned vocabulary is why this package ships no tokenizer:
// given the same normalized string, every platform produces the same bucket
// indices by construction.
//
// Iterated over **Unicode code points**, not UTF-16 code units. Iterating with a
// plain index would split astral characters into surrogate pairs and hash
// something the reference never sees.

const OFFSET_BASIS = 0x811c9dc5;
const PRIME = 0x01000193;

export const NGRAM_ORDERS = [1, 2, 3, 4, 5] as const;
export const offsetBasis = OFFSET_BASIS;
export const prime = PRIME;

/** FNV-1a 32-bit over Unicode code points. */
export function fnv1a(text: string): number {
  let hash = OFFSET_BASIS;
  for (const char of text) {
    hash ^= char.codePointAt(0)!;
    // Math.imul keeps the multiply in 32 bits, which `*` would not.
    hash = Math.imul(hash, PRIME) & 0xffffffff;
  }
  return hash >>> 0;
}

function fnv1aOfCodePoints(codePoints: number[], from: number, until: number): number {
  let hash = OFFSET_BASIS;
  for (let index = from; index < until; index++) {
    hash ^= codePoints[index]!;
    hash = Math.imul(hash, PRIME) & 0xffffffff;
  }
  return hash >>> 0;
}

/**
 * Bucket counts for a normalized string: the bag the model consumes.
 *
 * Each whitespace token is wrapped in `^`/`$` so word-initial and word-final
 * sequences stay distinguishable from word-internal ones. That distinction
 * carries much of the signal — Portuguese `ão$`, Italian `^gli`.
 */
export function buckets(
  text: string,
  numBuckets: number,
  orders: readonly number[] = NGRAM_ORDERS,
): Map<number, number> {
  const counts = new Map<number, number>();
  if (!text) return counts;
  for (const token of text.split(" ")) {
    if (!token) continue;
    const marked = [0x5e, ...[...token].map((c) => c.codePointAt(0)!), 0x24]; // ^ token $
    for (const order of orders) {
      if (order > marked.length) continue;
      for (let start = 0; start + order <= marked.length; start++) {
        const bucket = fnv1aOfCodePoints(marked, start, start + order) % numBuckets;
        counts.set(bucket, (counts.get(bucket) ?? 0) + 1);
      }
    }
  }
  return counts;
}

File: packages/tongue-kotlin/src/main/kotlin/ai/desertant/tongue/Hashing.kt
package ai.desertant.tongue

/**
 * FNV-1a character n-gram featurizer, ported from src/tongue_training/hashing.py.
 *
 * Hashing instead of a learned vocabulary is why this SDK ships no tokenizer:
 * given the same normalized string, every platform produces the same bucket
 * indices by construction, with no BPE merge-order drift to reconcile.
 *
 * Iterated over **Unicode code points** — not UTF-8 bytes and not UTF-16 `char`s.
 * Iterating `char` would split astral characters into surrogate pairs and hash
 * something the reference never sees.
 */
public object Hashing {
    public const val OFFSET_BASIS: Int = -0x7EE3623B     // 0x811C9DC5 as a signed Int
    public const val PRIME: Int = 0x01000193
    public val NGRAM_ORDERS: List<Int> = listOf(1, 2, 3, 4, 5)

    private const val BOUNDARY_START = '^'
    private const val BOUNDARY_END = '$'

    /** FNV-1a 32-bit over Unicode code points, returned as an unsigned Long. */
    public fun fnv1a(text: String): Long {
        var hash = OFFSET_BASIS
        var index = 0
        while (index < text.length) {
            val codePoint = text.codePointAt(index)
            hash = (hash xor codePoint) * PRIME
            index += Character.charCount(codePoint)
        }
        return hash.toLong() and 0xFFFFFFFFL
    }

    /** FNV-1a over a code-point slice, avoiding a substring per n-gram. */
    private fun fnv1a(codePoints: IntArray, from: Int, until: Int): Long {
        var hash = OFFSET_BASIS
        for (index in from until until) hash = (hash xor codePoints[index]) * PRIME
        return hash.toLong() and 0xFFFFFFFFL
    }

    /**
     * Bucket counts for a normalized string: the bag the model consumes.
     *
     * Each whitespace token is wrapped in `^`/`$` so word-initial and word-final
     * sequences stay distinguishable from word-internal ones. That distinction
     * carries much of the signal — Portuguese `ão$`, Italian `^gli`.
     */
    @JvmOverloads
    public fun buckets(
        text: String,
        numBuckets: Int,
        orders: List<Int> = NGRAM_ORDERS,
    ): Map<Int, Int> {
        if (text.isEmpty()) return emptyMap()
        val counts = HashMap<Int, Int>()
        for (token in text.split(" ")) {
            if (token.isEmpty()) continue
            val marked = buildMarked(token)
            for (order in orders) {
                if (order > marked.size) continue
                for (start in 0..(marked.size - order)) {
                    val bucket = (fnv1a(marked, start, start + order) % numBuckets).toInt()
                    counts[bucket] = (counts[bucket] ?: 0) + 1
                }
            }
        }
        return counts
    }

    private fun buildMarked(token: String): IntArray {
        val inner = token.codePoints().toArray()
        val marked = IntArray(inner.size + 2)
        marked[0] = BOUNDARY_START.code
        inner.copyInto(marked, 1)
        marked[marked.size - 1] = BOUNDARY_END.code
        return marked
    }
}
